File: ambeth/Ambeth.Cache.Bytecode/ambeth/ioc/CacheBytecodeModule.cs
using De.Osthus.Ambeth.Bytecode.Behavior;
using De.Osthus.Ambeth.Databinding;
using De.Osthus.Ambeth.Ioc.Annotation;
using De.Osthus.Ambeth.Ioc.Factory;
using De.Osthus.Ambeth.Template;

namespace De.Osthus.Ambeth.Ioc
{
    [FrameworkModule]
    public class CacheBytecodeModule : IInitializingModule
    {
        public void AfterPropertiesSet(IBeanContextFactory beanContextFactory)
        {
            // cascade $1
            BytecodeModule.AddDefaultBytecodeBehavior<EmbeddedTypeBehavior>(beanContextFactory);
            BytecodeModule.AddDefaultBytecodeBehavior<EnhancedTypeBehavior>(beanContextFactory);
            BytecodeModule.AddDefaultBytecodeBehavior<DefaultPropertiesBehavior>(beanContextFactory);
            // cascade $2
            BytecodeModule.AddDefaultBytecodeBehavior<LazyRelationsBehavior>(beanContextFactory);

            BytecodeModule.AddDefaultBytecodeBehavior<InitializeEmbeddedMemberBehavior>(beanContextFactory);
            // cascade $3
            BytecodeModule.AddDefaultBytecodeBehavior<NotifyPropertyChangedBehavior>(beanContextFactory);
            BytecodeModule.AddDefaultBytecodeBehavior<ParentCacheHardRefBehavior>(beanContextFactory);
            BytecodeModule.AddDefaultBytecodeBehavior<EntityEqualsBehavior>(beanContextFactory);
            //BytecodeModule.AddDefaultBytecodeBehavior<PublicEmbeddedConstructorBehavior>(beanContextFactory);
            // cascade $4
            BytecodeModule.AddDefaultBytecodeBehavior<DataObjectBehavior>(beanContextFactory);

            BytecodeModule.AddDefaultBytecodeBehavior<CacheMapEntryBehavior>(beanContextFactory);
            BytecodeModule.AddDefaultBytecodeBehavior<RootCacheValueBehavior>(beanContextFactory);

            beanContextFactory.RegisterAnonymousBean<DataObjectTemplate>().Autowireable<DataObjectTemplate>();
            beanContextFactory.RegisterAnonymousBean<EntityEqualsTemplate>().Autowireable<EntityEqualsTemplate>();
            beanContextFactory.RegisterAnonymousBean<EmbeddedTypeTemplate>().Autowireable<EmbeddedTypeTemplate>();
            beanContextFactory.RegisterAnonymousBean<PropertyChangeTemplate>().Autowireable(typeof(PropertyChangeTemplate),
                typeof(IPropertyChangeExtensionExtendable), typeof(ICollectionChangeExtensionExtendable));
            beanContextFactory.RegisterAnonymousBean<ValueHolderContainerTemplate>().Autowireable<ValueHolderContainerTemplate>();
        }
    }
}

File: jambeth/jambeth-cache-bytecode/src/main/java/de/osthus/ambeth/ioc/CacheBytecodeModule.java
package de.osthus.ambeth.ioc;

import de.osthus.ambeth.bytecode.behavior.CacheMapEntryBehavior;
import de.osthus.ambeth.bytecode.behavior.DataObjectBehavior;
import de.osthus.ambeth.bytecode.behavior.DefaultPropertiesBehavior;
import de.osthus.ambeth.bytecode.behavior.EmbeddedTypeBehavior;
import de.osthus.ambeth.bytecode.behavior.EnhancedTypeBehavior;
import de.osthus.ambeth.bytecode.behavior.EntityEqualsBehavior;
import de.osthus.ambeth.bytecode.behavior.InitializeEmbeddedMemberBehavior;
import de.osthus.ambeth.bytecode.behavior.LazyRelationsBehavior;
import de.osthus.ambeth.bytecode.behavior.NotifyPropertyChangedBehavior;
import de.osthus.ambeth.bytecode.behavior.ParentCacheHardRefBehavior;
import de.osthus.ambeth.bytecode.behavior.RootCacheValueBehavior;
import de.osthus.ambeth.ioc.annotation.FrameworkModule;
import de.osthus.ambeth.ioc.factory.IBeanContextFactory;

@FrameworkModule
public class CacheBytecodeModule implements IInitializingModule
{
	@Override
	public void afterPropertiesSet(IBeanContextFactory beanContextFactory) throws Throwable
	{
		// cascade $1
		BytecodeModule.addDefaultBytecodeBehavior(beanContextFactory, EmbeddedTypeBehavior.class);
		BytecodeModule.addDefaultBytecodeBehavior(beanContextFactory, EnhancedTypeBehavior.class);
		BytecodeModule.addDefaultBytecodeBehavior(beanContextFactory, DefaultPropertiesBehavior.class);
		// cascade $2
		BytecodeModule.addDefaultBytecodeBehavior(beanContextFactory, LazyRelationsBehavior.class);

		BytecodeModule.addDefaultBytecodeBehavior(beanContextFactory, InitializeEmbeddedMemberBehavior.class);
		// cascade $3
		BytecodeModule.addDefaultBytecodeBehavior(beanContextFactory, NotifyPropertyChangedBehavior.class);
		BytecodeModule.addDefaultBytecodeBehavior(beanContextFactory, ParentCacheHardRefBehavior.class);
		BytecodeModule.addDefaultBytecodeBehavior(beanContextFactory, EntityEqualsBehavior.class);
		// BytecodeModule.addDefaultBytecodeBehavior(beanContextFactory, PublicEmbeddedConstructorBehavior.class);
		// cascade $4
		BytecodeModule.addDefaultBytecodeBehavior(beanContextFactory, DataObjectBehavior.class);

		BytecodeModule.addDefaultBytecodeBehavior(beanContextFactory, CacheMapEntryBehavior.class);
		BytecodeModule.addDefaultBytecodeBehavior(beanContextFactory, RootCacheValueBehavior.class);
	}
}
